core/src/test/scala/org/apache/spark/ui/memory/MemoryListenerSuite.scala

```diff
@@ -0,0 +1,258 @@
+    listener.onStageSubmitted(MemoryListenerSuite.createStageStartEvent(3))
+    listener.onStageCompleted(MemoryListenerSuite.createStageEndEvent(3))
+
+    // both stage 2 and stage 3 will use the metrics last updated in stage 1
+    val mapForStage2 = listener.completedStagesToMem.get((2, 0)).get
+    val memInfo2 = mapForStage2.get(execId1).get
+    assert(memInfo2.transportInfo.isDefined)
+    val transMetrics2 = memInfo2.transportInfo.get
+    assert((15, 15, MemTime(15, 0), MemTime(15, 0)) === (transMetrics2.onHeapSize,
+      transMetrics2.offHeapSize,
+      transMetrics2.peakOnHeapSizeTime,
+      transMetrics2.peakOffHeapSizeTime))
+
+    val mapForStage3 = listener.completedStagesToMem.get((3, 0)).get
+    val memInfo3 = mapForStage3.get(execId1).get
+    assert(memInfo3.transportInfo.isDefined)
+    val transMetrics3 = memInfo3.transportInfo.get
+    assert((15, 15, MemTime(15, 0), MemTime(15, 0)) === (transMetrics3.onHeapSize,
+      transMetrics3.offHeapSize,
+      transMetrics3.peakOnHeapSizeTime,
+      transMetrics3.peakOffHeapSizeTime))
+
+    listener.onExecutorRemoved(SparkListenerExecutorRemoved(0L, execId1, ""))
+  }
+
+  test("test multiple executors") {
+    val listener = new MemoryListener
+    val execId1 = "exec-1"
+    val execId2 = "exec-2"
+    val execId3 = "exec-3"
+
+    // two executors added first
+    listener.onExecutorAdded(
+      SparkListenerExecutorAdded(0L, execId1, new ExecutorInfo("host1", 1, Map.empty)))
+    listener.onExecutorAdded(
+      SparkListenerExecutorAdded(0L, execId2, new ExecutorInfo("host2", 1, Map.empty)))
+
+    // three executors running in one stage and one executor is removed before stage complete
+    listener.onStageSubmitted(MemoryListenerSuite.createStageStartEvent(1))
+    val exec1Metrics = MemoryListenerSuite.createExecutorMetrics("host-1", 1446336000L, 20, 10)
+    listener.onExecutorMetricsUpdate(MemoryListenerSuite.createExecutorMetricsUpdateEvent(
+      execId1, exec1Metrics))
+    val exec2Metrics = MemoryListenerSuite.createExecutorMetrics("host-2", 1446337000L, 15, 5)
+    listener.onExecutorMetricsUpdate(MemoryListenerSuite.createExecutorMetricsUpdateEvent(
+      execId2, exec2Metrics))
+    // one more executor added during the stage is running
+    listener.onExecutorAdded(
+      SparkListenerExecutorAdded(0L, execId3, new ExecutorInfo("host3", 1, Map.empty)))
+    val exec3Metrics = MemoryListenerSuite.createExecutorMetrics("host-3", 1446338000L, 30, 15)
+    listener.onExecutorMetricsUpdate(MemoryListenerSuite.createExecutorMetricsUpdateEvent(
+      execId3, exec3Metrics))
+    // executor 2 removed before stage complete
+    listener.onExecutorRemoved(SparkListenerExecutorRemoved(0L, execId2, ""))
+    listener.onStageCompleted(MemoryListenerSuite.createStageEndEvent(1))
+
+    listener.onExecutorRemoved(SparkListenerExecutorRemoved(0L, execId1, ""))
+    listener.onExecutorRemoved(SparkListenerExecutorRemoved(0L, execId3, ""))
```

Comment:
these times should be consistent w/ the ExecutorMetric times.  In fact, maybe you also need to supply times to createStageStart / End.  Note that the times can also be really simple -- 1,2,3, ...